### PR TITLE
update for null safety

### PR DESCRIPTION
**Pull request by devoncarew**
- update to support 2.12.0 and null safety

### CODE REVIEW

File: lib/src/usage_impl_io.dart [hunk 1]
```diff
@@ -87,11 +87,11 @@ class IOPostHandler extends PostHandler {
 
     if (_client == null) {
       _client = mockClient ?? HttpClient();
-      _client.userAgent = _userAgent;
+      _client!.userAgent = _userAgent;
```
Comment (pq) [inline lib/src/usage_impl_io.dart:90]: Seems odd to need the bang right after an assignment to `_client`... Isn't it sure to be non-null here?

File: lib/src/usage_impl_io.dart [hunk 2]
```diff
@@ -108,10 +108,10 @@ class IOPostHandler extends PostHandler {
...
-  File _file;
-  Map _map;
+  late File _file;
+  Map? _map;
 
-  IOPersistentProperties(String name, {String documentDirPath}) : super(name) {
+  IOPersistentProperties(String name, {String? documentDirPath}) : super(name) {
     var fileName = '.${name.replaceAll(' ', '_')}';
     documentDirPath ??= userHomeDir();
     _file = File(path.join(documentDirPath, fileName));
```
Comment (pq) [inline lib/src/usage_impl_io.dart:117]: Again, with this assignment, I'm surprised that the bang is required below.

Ah OK.  Fields don't promote (a la https://dart.dev/null-safety/understanding-null-safety).  Disregard my comments above...

File: lib/src/usage_impl_io.dart [hunk 3]
```diff
@@ -130,17 +130,17 @@ class IOPersistentProperties extends PersistentProperties {
...
 
   @override
-  dynamic operator [](String key) => _map[key];
+  dynamic operator [](String key) => _map![key];
 
   @override
   void operator []=(String key, dynamic value) {
-    if (value == null && !_map.containsKey(key)) return;
-    if (_map[key] == value) return;
+    if (value == null && !_map!.containsKey(key)) return;
```
Comment (pq) [inline lib/src/usage_impl_io.dart:137]: I'd be inclined to mark `_map` late so you can drop all of these bangs.

File: pubspec.yaml [hunk 4]
```diff
@@ -3,15 +3,15 @@
...
 description: A Google Analytics wrapper for command-line, web, and Flutter apps.
 homepage: https://github.com/dart-lang/usage
 
 environment:
-  sdk: '>=2.2.0 <3.0.0'
+  sdk: '>=2.12.0-0 <3.0.0'
 
 dependencies:
-  path: ^1.4.0
+  path: ^1.8.0-nullsafety.3
```
Comment (pq) [inline pubspec.yaml:14]: Curious: why the `.3` lower bound?

**Review by pq — COMMENTED** [review]
I feel like I'm just figuring this out as I go...  Hope I'm not just muddying the waters with all the questions!

File: example/example.dart [hunk 5]
```diff
@@ -9,40 +9,40 @@ import 'dart:html';
...
+  querySelector('#foo')!.onClick.listen((_) => _handleFoo());
+  querySelector('#bar')!.onClick.listen((_) => _handleBar());
+  querySelector('#page')!.onClick.listen((_) => _changePage());
 }
 
-String _ua() => (querySelector('#ua') as InputElement).value.trim();
+String _ua() => (querySelector('#ua') as InputElement).value!.trim();
 
-Analytics getAnalytics() {
+Analytics? getAnalytics() {
```
[the remark this answers is not in the document]
Comment (devoncarew) [reply example/example.dart]: Good point - done.

File: example/example.dart [hunk 6]
```diff
@@ -9,40 +9,40 @@ import 'dart:html';
...
+    _analytics = AnalyticsHtml(_lastUa!, 'Test app', '1.0');
+    _analytics!.sendScreenView(window.location.pathname!);
   }
 
   return _analytics;
 }
 
 void _handleFoo() {
-  var analytics = getAnalytics();
+  var analytics = getAnalytics()!;
```
[the remark this answers is not in the document]
Comment (devoncarew) [reply example/example.dart]: Yup - removed.

File: lib/src/usage_impl.dart [hunk 7]
```diff
@@ -24,8 +24,8 @@ String postEncode(Map<String, dynamic> map) {
 /// overall average rate of operations to 1 per second.
 class ThrottlingBucket {
   final int startingCount;
-  int drops;
-  int _lastReplenish;
+  late int drops;
+  late int _lastReplenish;
 
   ThrottlingBucket(this.startingCount) {
```
[the remark this answers is not in the document]
Comment (devoncarew) [reply lib/src/usage_impl.dart]: 👍 - yup, this works

File: lib/src/usage_impl_io.dart (hunk 1 above)
Comment (devoncarew) [reply lib/src/usage_impl_io.dart:90]: yup, the fields-don't-promote thing

File: lib/src/usage_impl_io.dart (hunk 3 above)
Comment (devoncarew) [reply lib/src/usage_impl_io.dart]: done

File: pubspec.yaml (hunk 4 above)
Comment (devoncarew) [reply pubspec.yaml]: This was just the latest available - I can drop to `1.8.0-nullsafety`.

**Comment by devoncarew** [comment]
Thanks for the review! Updated with the latest - ptal.

**Review by pq — APPROVED** [review]
👍

**Pull request merged**